src/workerd/api/headers.c++

```diff
@@ -1,134 +1,367 @@
+        (static_cast<uint64_t>(static_cast<uint8_t>(a[i + 2])) << 16) |
+        (static_cast<uint64_t>(static_cast<uint8_t>(a[i + 3])) << 24) |
+        (static_cast<uint64_t>(static_cast<uint8_t>(a[i + 4])) << 32) |
+        (static_cast<uint64_t>(static_cast<uint8_t>(a[i + 5])) << 40) |
+        (static_cast<uint64_t>(static_cast<uint8_t>(a[i + 6])) << 48) |
+        (static_cast<uint64_t>(static_cast<uint8_t>(a[i + 7])) << 56);
+
+    uint8_t b0 = b[i + 0];
+    uint8_t b1 = b[i + 1];
+    uint8_t b2 = b[i + 2];
+    uint8_t b3 = b[i + 3];
+    uint8_t b4 = b[i + 4];
+    uint8_t b5 = b[i + 5];
+    uint8_t b6 = b[i + 6];
+    uint8_t b7 = b[i + 7];
+
+    uint64_t normalized_b = (static_cast<uint64_t>(b0 + CASE_CONVERSION_TABLE[b0]) << 0) |
+        (static_cast<uint64_t>(b1 + CASE_CONVERSION_TABLE[b1]) << 8) |
+        (static_cast<uint64_t>(b2 + CASE_CONVERSION_TABLE[b2]) << 16) |
+        (static_cast<uint64_t>(b3 + CASE_CONVERSION_TABLE[b3]) << 24) |
+        (static_cast<uint64_t>(b4 + CASE_CONVERSION_TABLE[b4]) << 32) |
+        (static_cast<uint64_t>(b5 + CASE_CONVERSION_TABLE[b5]) << 40) |
+        (static_cast<uint64_t>(b6 + CASE_CONVERSION_TABLE[b6]) << 48) |
+        (static_cast<uint64_t>(b7 + CASE_CONVERSION_TABLE[b7]) << 56);
+
+    if (wa != normalized_b) return false;
+    i += sizeof(uint64_t);
+  }
+
+  for (; i < a.size(); ++i) {
+    char cb = b[i] + CASE_CONVERSION_TABLE[static_cast<kj::byte>(b[i])];
+    if (a[i] != cb) return false;
+  }
+  return true;
+}
+
+constexpr uint caseInsensitiveHash(kj::StringPtr name) {
+  static constexpr uint32_t MULTIPLIER = 16777619u;
+  uint hash = 2166136261u;
+  size_t i = 0;
+
+  while (i + sizeof(uint64_t) <= name.size()) {
+    uint8_t b0 = name[i + 0];
+    uint8_t b1 = name[i + 1];
+    uint8_t b2 = name[i + 2];
+    uint8_t b3 = name[i + 3];
+    uint8_t b4 = name[i + 4];
+    uint8_t b5 = name[i + 5];
+    uint8_t b6 = name[i + 6];
+    uint8_t b7 = name[i + 7];
+
+    hash ^= b0 + CASE_CONVERSION_TABLE[b0];
+    hash *= MULTIPLIER;
+    hash ^= b1 + CASE_CONVERSION_TABLE[b1];
+    hash *= MULTIPLIER;
+    hash ^= b2 + CASE_CONVERSION_TABLE[b2];
+    hash *= MULTIPLIER;
+    hash ^= b3 + CASE_CONVERSION_TABLE[b3];
+    hash *= MULTIPLIER;
+    hash ^= b4 + CASE_CONVERSION_TABLE[b4];
+    hash *= MULTIPLIER;
+    hash ^= b5 + CASE_CONVERSION_TABLE[b5];
+    hash *= MULTIPLIER;
+    hash ^= b6 + CASE_CONVERSION_TABLE[b6];
+    hash *= MULTIPLIER;
+    hash ^= b7 + CASE_CONVERSION_TABLE[b7];
+    hash *= MULTIPLIER;
+
+    i += sizeof(uint64_t);
+  }
+
+  // Handle remaining bytes
+  for (; i < name.size(); ++i) {
+    uint8_t b = name[i];
+    hash ^= b + CASE_CONVERSION_TABLE[b];
+    hash *= MULTIPLIER;
+  }
+
+  hash = (hash >> 16) ^ hash;
+  return hash;
+}
+
+constexpr size_t HEADER_MAP_SIZE = 512;
+
+struct HeaderHashTable final {
+  struct Entry {
+    kj::StringPtr name;
+    uint id;
+  };
+
+  Entry entries[HEADER_MAP_SIZE] = {};
+  size_t kMaxEntrySize = 0;
```

Comment:
Nit: this reads kinda weird, since `k` is often a marker of a constant. Maybe just a `// Calculated in constructor` comment?